Add PortableTest to test nested portable serializer

ubeydk opened the pull request:
Related unit test is added, and test is passed without any code modification. It seems that there is no problem.

Related classes and fields has been compared to make sure that implementation is as expected. 
- Related properties are already private/readonly as suggested [here](https://github.com/hazelcast/hazelcast/pull/12734) . 
- By the way, there is no **version** property in FieldDefinition class but there is in java equivalent. I intended to add it but it seems it is not mandatory. I can add if you think it is necessary.

@asimarslan 
**Fixes #163**

asimarslan commented:
verify

devOpsHazelcast commented:
Linux Net Core Tests PASSed.

asimarslan commented:
Hi @ubeydk ,
Thanks for the PR. In order to merge your pull request, we need the contributor to sign [Hazelcast Contributor Agreement](https://hazelcast.atlassian.net/wiki/display/COM/Hazelcast+Contributor+Agreement).
Can you please send the contributor agreement?

sancar commented:
Test is passing because there is a missing implementation as far as I understand. 
Test fails with the fix https://github.com/hazelcast/hazelcast/pull/9415
Which means the missing part that you see `version property in FieldDefinition` is actually necessary for this test to fail. 
I would advise to first apply the fix above, 
see the test fail , 
then apply this fix https://github.com/hazelcast/hazelcast/pull/12735/files#diff-7656a5c995c98501c8683df049f675e5R123

devOpsHazelcast commented:
Windows .Net Framework Tests  PASSed.

devOpsHazelcast commented:
Windows Net Core Tests PASSed.

sancar commented:
It looks like not all changes are reflected. 
For example default value of version in class definition builder should change. https://github.com/hazelcast/hazelcast-csharp-client/pull/179/files#diff-d942257f2c7076f0dc4de937b61d4742L46

Can you also add the test `NestedPredicateVersionedPortablesTest.java` https://github.com/hazelcast/hazelcast/pull/9415/files#diff-78ff247a43e5f9a17e48aea10682ac8a
 from https://github.com/hazelcast/hazelcast/pull/9415

ubeydk commented:
Hi @asimarslan ,
I have sent contributor agreement.

Thanks for your review @sancar . It is my big mistake that didn't realize [it](https://github.com/hazelcast/hazelcast/pull/9415/files) is complete java implementation of adding version property to FieldDefinition. So I have tried to add it by fixing compilation errors and resolving test fails on my own. I will try to reflect all missing changes [here](https://github.com/hazelcast/hazelcast/pull/9415/files).

ubeydk commented:
There are still missing changes in 2 files.
 - [PortablePositionFactoryTest](https://github.com/hazelcast/hazelcast/pull/9415/files#diff-4d695aa6a719335744e9d51d1a3d9b49): It seems that there is no implementation of PortablePosition (interface and related classes: navigator, factory and test).


 -  [NestedPredicateVersionedPortablesTest](https://github.com/hazelcast/hazelcast/pull/9415/files#diff-78ff247a43e5f9a17e48aea10682ac8a): I got errors from tests which require initiated HazelcastInstance. So, even If I create this file, I won't be able to run tests properly. I am getting these errors from various tests:
1. `Thrift.Transport.TTransportException : Cannot read, Remote side has closed`
2. `System.Net.Sockets.SocketException: No connection could be made because the target machine actively refused it 127.0.0.1:9701`

I thought remote-controller is not started properly. Then I started remote-controller manually and got these errors from remote-controller snapshot console.
`ERROR [AbstractNonblockingServer$FrameBuffer] - Unexpected throwable while invoking!
java.lang.NoClassDefFoundError: com/hazelcast/config/XmlConfigBuilder
        at com.hazelcast.remotecontroller.HzCluster.<init>(HzCluster.java:32)`

`ERROR [TNonblockingServer] - run() exiting due to uncaught error
java.lang.NoClassDefFoundError: com/hazelcast/core/Hazelcast
        at com.hazelcast.remotecontroller.ClusterManager.shutdownAll(ClusterManager.java:114)`

By the way, I have modified build.bat file by changing 1 line (117th) that starts unit tests. "build.bat" and "console-text.xml" files have been [**attached**](https://github.com/hazelcast/hazelcast-csharp-client/files/2338326/hazelcast-csharp-client.zip).



.

sancar commented:
`PortablePositionFactoryTest` is part of another feature. No need to worry about that on this pr. 

For `NestedPredicateVersionedPortablesTest:
Tests themselves starts the remote controller. Check the other tests. For example:
https://github.com/hazelcast/hazelcast-csharp-client/blob/master/Hazelcast.Test/Hazelcast.Client.Test/ClientMapTest.cs
This class extends the https://github.com/hazelcast/hazelcast-csharp-client/blob/master/Hazelcast.Test/Hazelcast.Client.Test/SingleMemberBaseTest.cs
and that is were remote controller started. 

Your test should have similar structure.

ubeydk commented:
Actually, I mean tests are not working properly on my local computer. I added [NestedPredicateVersionedPortablesTest](https://github.com/hazelcast/hazelcast-csharp-client/pull/179/files#diff-8db9a169b831250e2b955b6694c95044) but didn't be able to properly test it. Small edits may be needed.

In addition, I added and then commented out `singleAttributeQuery_versionedProtables_predicates()` test, because I did not find equivalent of **EntryObject** and **PredicateBuilder** class.

sancar commented:
Can you properly run https://github.com/hazelcast/hazelcast-csharp-client/blob/master/Hazelcast.Test/Hazelcast.Client.Test/ClientMapTest.cs ? 
If you can, I would recommend you to take a look at the difference and what is missing from your implementation. 
If you can not you should take a look at to build.bat in detail. 
`build.bat` should be able to download and start necessary projects. You don't need to start remote controller yourself. But make sure that each step works correctly in build.bat.

ubeydk commented:
Finally, I successfully run the tests, and all the new changes' tests are passed. 

It seems everything works.

sancar commented:
run-lab-run

asimarslan commented:
verify

devOpsHazelcast commented:
Linux Net Core Tests PASSed.

devOpsHazelcast commented:
Windows .Net Framework Tests  PASSed.

devOpsHazelcast commented:
Windows Net Core Tests PASSed.

asimarslan commented on Hazelcast.Test/Hazelcast.Client.Test/NestedPredicateVersionedPortablesTest.cs:
```diff
@@ -0,0 +1,251 @@
+        
+        /* TODO
+        [Test]
+        public void singleAttributeQuery_versionedProtables_predicates()
```
Are we missing a test or is it left over code?

ubeydk replied:
In this test, EntryObject and PredicateBuilder classes are used but there are no equivalent classes in C#. This test should be added when related classes are implemented.

Actually, I was not sure about what should I do, then I left it as commented out for your reviews.

asimarslan replied:
You can use predicate as :
```
var predicate = Predicates.Property("limb.name").Equal("hand");
```

asimarslan commented on Hazelcast.Test/Hazelcast.Client.Test.Serialization/PortableTest.cs:
```diff
@@ -0,0 +1,164 @@
+{
+    internal class PortableTest
+    {
+        // Test for issue https://github.com/hazelcast/hazelcast/issues/12733
```
we don't need this comment as this is .net fix. You can delete the comment

ubeydk replied:
Okay, It will be deleted in next commit

asimarslan commented:
verify

devOpsHazelcast commented:
Linux Net Core Tests PASSed.

devOpsHazelcast commented:
Windows .Net Framework Tests  PASSed.

devOpsHazelcast commented:
Windows Net Core Tests PASSed.
